### PR TITLE
Build better snowflake exception when PK is missing

### PR DESCRIPTION
**Pull request by evantahler**
We still want snowflake to fail T&D if there are records missing PKs, but this PR makes the error less obtuse.

Re: https://github.com/airbytehq/oncall/issues/2842

### CODE REVIEW

File: airbyte-integrations/connectors/destination-snowflake/src/test-integration/java/io/airbyte/integrations/destination/snowflake/typing_deduping/SnowflakeSqlGeneratorIntegrationTest.java [hunk 1]
```diff
@@ -369,4 +372,24 @@ record -> record.get(JavaBaseConstants.COLUMN_NAME_AB_RAW_ID).asText(),
...
+        streamId,
+        BaseTypingDedupingTest.readRecords("sqlgenerator/incrementaldedup_inputrecords_missing_pk.jsonl"));
+
+    final String sql = generator.updateTable(incrementalDedupStream, "");
+    Exception exception = assertThrows(Exception.class, () -> { // TODO (evan): I don't really know what class of error the Snowflake driver throws
+      destinationHandler.execute(sql);
+    });
+
+    assertTrue(exception.getMessage().contains("_AB_MISSING_PRIMARY_KEY"));
+    assertTrue(exception.getMessage().contains("\"users_raw\"` has rows missing a primary key"));
```
Comment (edgao) [inline airbyte-integrations/connectors/destination-snowflake/src/test-integration/java/io/airbyte/integrations/destination/snowflake/typing_deduping/SnowflakeSqlGeneratorIntegrationTest.java:392]: this is sort of redundant with https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/bases/base-typing-deduping-test/src/main/java/io/airbyte/integrations/base/destination/typing_deduping/BaseSqlGeneratorIntegrationTest.java#L334

(your new test adds asserts on the specific error message though, which is probably valuable? maybe makes sense to actually override the existing test case)

File: airbyte-integrations/bases/base-typing-deduping-test/src/main/resources/sqlgenerator/incrementaldedup_inputrecords_missing_pk.jsonl [hunk 2]
```diff
@@ -0,0 +1,2 @@
+{"_airbyte_raw_id": "d7b81af0-01da-4846-a650-cc398986bc91", "_airbyte_extracted_at": "2023-01-01T00:00:00Z", "_airbyte_data": {"id1": 1, "id2": 100, "updated_at": "2023-01-01T01:00:00Z", "string": "Alice", "struct": {"city": "San Francisco", "state": "CA"}, "integer": 42}}
```
Comment (edgao) [inline airbyte-integrations/bases/base-typing-deduping-test/src/main/resources/sqlgenerator/incrementaldedup_inputrecords_missing_pk.jsonl:1]: this test is defined in snowflake, so I think this file should live in destination-snowflake

File: airbyte-integrations/connectors/destination-snowflake/src/test-integration/java/io/airbyte/integrations/destination/snowflake/typing_deduping/SnowflakeSqlGeneratorIntegrationTest.java [hunk 3]
```diff
@@ -369,4 +372,24 @@ record -> record.get(JavaBaseConstants.COLUMN_NAME_AB_RAW_ID).asText(),
...
+  @Test
+  protected void customExceptionWhenMissingPK() throws Exception {
+    createRawTable(streamId);
+    createFinalTable(incrementalDedupStream, "");
+    insertRawTableRecords(
+        streamId,
+        BaseTypingDedupingTest.readRecords("sqlgenerator/incrementaldedup_inputrecords_missing_pk.jsonl"));
+
+    final String sql = generator.updateTable(incrementalDedupStream, "");
+    Exception exception = assertThrows(Exception.class, () -> { // TODO (evan): I don't really know what class of error the Snowflake driver throws
```
Comment (edgao) [inline airbyte-integrations/connectors/destination-snowflake/src/test-integration/java/io/airbyte/integrations/destination/snowflake/typing_deduping/SnowflakeSqlGeneratorIntegrationTest.java:387]: run it in a debugger and see? (I can do this if your intellij is sad, it's probably a jdbc exception)

**Review by edgao — APPROVED** [review]
lgtm, just a few nits

**Review by edgao — APPROVED** [review]
:shipit:

**Comment by evantahler** [comment]
The `noCrashOnSpecialCharacters()` test is legit! Fixing now

File: airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/typing_deduping/SnowflakeSqlGenerator.java [hunk 4]
```diff
@@ -268,22 +268,24 @@ String validatePrimaryKeys(final StreamId id,
 
     final String script = new StringSubstitutor(Map.of(
         "raw_table_id", id.rawTableId(QUOTE),
+        "raw_table_id_no_single_quotes", id.rawTableId(QUOTE).replaceAll("'", ""),
```
Comment (edgao) [inline airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/typing_deduping/SnowflakeSqlGenerator.java:271]: `escapeSingleQuotedString` exists for this purpose! it also handles backslashes.

(... which is maybe a test case that the noCrash test should include :/ )

**Review by edgao — APPROVED** [review]

**Pull request merged**